### PR TITLE
Fix false detections of equi-joins

### PR DESCRIPTION
**Pull request by jeeminso**
## Summary of the changes / Why this improves CrateDB
~~Fixes https://github.com/crate/crate/issues/15613.~~ Follows https://github.com/crate/crate/pull/15682. Basically this fixes a variant of https://github.com/crate/crate/issues/15613:

```
SELECT * FROM t1 INNER  JOIN t0 ON (t1.c1 = t1.c1 and CASE 'q' WHEN t1.c0 THEN false ELSE (t0.c0 IN (t0.c0)) END);
```

There are two parts to this PR(the current commit is part1 and 2 not yet implemented).

---
**Part 1**

Before fix:
```
cr> explain verbose SELECT * FROM t1 INNER  JOIN t0 ON (t1.c1 = t1.c1) WHERE (CASE 'q' WHEN t1.c0 THEN false ELSE (t0.c0 IN (t0.c0)) END );  
                                                                                                                                                                                              
+-----------------------------+----------------------------------------------------------------------------------------------------+
| STEP                        | QUERY PLAN                                                                                         |
+-----------------------------+----------------------------------------------------------------------------------------------------+
| Initial logical plan        | Join[INNER | ((c1 = c1) AND case(true, (c0 = ANY(_array(c0))), ('q' = c0), false))] (rows=unknown) |
|                             |   ├ Collect[doc.t1 | [c0, c1] | true] (rows=unknown)                                               |
|                             |   └ Collect[doc.t0 | [c0] | true] (rows=unknown)                                                   |
| optimizer_rewrite_join_plan | HashJoin[((c1 = c1) AND case(true, (c0 = ANY(_array(c0))), ('q' = c0), false))] (rows=unknown)     |
|                             |   ├ Collect[doc.t1 | [c0, c1] | true] (rows=unknown)                                               |
|                             |   └ Collect[doc.t0 | [c0] | true] (rows=unknown)                                                   |
| Final logical plan          | HashJoin[((c1 = c1) AND case(true, (c0 = ANY(_array(c0))), ('q' = c0), false))] (rows=unknown)     |
|                             |   ├ Collect[doc.t1 | [c0, c1] | true] (rows=unknown)                                               |
|                             |   └ Collect[doc.t0 | [c0] | true] (rows=unknown)                                                   |
+-----------------------------+----------------------------------------------------------------------------------------------------+
```
After:
```
cr> explain verbose SELECT * FROM t1 INNER  JOIN t0 ON (t1.c1 = t1.c1) WHERE (CASE 'q' WHEN t1.c0 THEN false ELSE (t0.c0 IN (t0.c0)) END );  
                                                                                                                                                                                              
+-------------------------------------------------------------+--------------------------------------------------------------------------------------------------------------+
| STEP                                                        | QUERY PLAN                                                                                                   |
+-------------------------------------------------------------+--------------------------------------------------------------------------------------------------------------+
| Initial logical plan                                        | Join[INNER | ((c1 = c1) AND case(true, (c0 = ANY(_array(c0))), ('q' = c0), false))] (rows=unknown)           |
|                                                             |   ├ Collect[doc.t1 | [c0, c1] | true] (rows=unknown)                                                         |
|                                                             |   └ Collect[doc.t0 | [c0] | true] (rows=unknown)                                                             |
| optimizer_rewrite_join_plan                                 | NestedLoopJoin[INNER | ((c1 = c1) AND case(true, (c0 = ANY(_array(c0))), ('q' = c0), false))] (rows=unknown) |
|                                                             |   ├ Collect[doc.t1 | [c0, c1] | true] (rows=unknown)                                                         |
|                                                             |   └ Collect[doc.t0 | [c0] | true] (rows=unknown)                                                             |
| optimizer_move_constant_join_conditions_beneath_nested_loop | HashJoin[case(true, (c0 = ANY(_array(c0))), ('q' = c0), false)] (rows=unknown)                               |
|                                                             |   ├ Filter[(c1 = c1)] (rows=0)                                                                               |
|                                                             |   │  └ Collect[doc.t1 | [c0, c1] | true] (rows=unknown)                                                      |
|                                                             |   └ Collect[doc.t0 | [c0] | true] (rows=unknown)                                                             |
| optimizer_merge_filter_and_collect                          | HashJoin[case(true, (c0 = ANY(_array(c0))), ('q' = c0), false)] (rows=unknown)                               |
|                                                             |   ├ Collect[doc.t1 | [c0, c1] | (c1 = c1)] (rows=unknown)                                                    |
|                                                             |   └ Collect[doc.t0 | [c0] | true] (rows=unknown)                                                             |
| Final logical plan                                          | HashJoin[case(true, (c0 = ANY(_array(c0))), ('q' = c0), false)] (rows=unknown)                               |
|                                                             |   ├ Collect[doc.t1 | [c0, c1] | (NOT (c1 IS NULL))] (rows=unknown)                                           |
|                                                             |   └ Collect[doc.t0 | [c0] | true] (rows=unknown)                                                             |
+-------------------------------------------------------------+--------------------------------------------------------------------------------------------------------------+
```

=> Basically fixed `EquiJoinDetector.isHashJoinPossible()` to return `false` for the bug scenario but it eventually becomes a hash join. A hash join with such join condition is invalid. `optimizer_move_constant_join_conditions_beneath_nested_loop` should be able to move the constant condition down without rewriting it to a hash join.

---
**Part 2**

This fix causes test failures which seem to be related to `optimizer_move_constant_join_conditions_beneath_nested_loop`, for example:
```
--before:
cr> explain verbose select doc.t1.*, doc.t2.b from doc.t1 join doc.t2 on doc.t1.x = doc.t2.y and doc.t2.b = 'abc';                                                                            
+-------------------------------------------------------------+----------------------------------------------------------------------+
| STEP                                                        | QUERY PLAN                                                           |
+-------------------------------------------------------------+----------------------------------------------------------------------+
| Initial logical plan                                        | Eval[x, b] (rows=unknown)                                            |
|                                                             |   └ Join[INNER | ((x = y) AND (b = 'abc'))] (rows=unknown)           |
|                                                             |     ├ Collect[doc.t1 | [x] | true] (rows=unknown)                    |
|                                                             |     └ Collect[doc.t2 | [b, y] | true] (rows=unknown)                 |
| optimizer_rewrite_join_plan                                 | Eval[x, b] (rows=unknown)                                            |
|                                                             |   └ NestedLoopJoin[INNER | ((x = y) AND (b = 'abc'))] (rows=unknown) |
|                                                             |     ├ Collect[doc.t1 | [x] | true] (rows=unknown)                    |
|                                                             |     └ Collect[doc.t2 | [b, y] | true] (rows=unknown)                 |
| optimizer_move_constant_join_conditions_beneath_nested_loop | Eval[x, b] (rows=unknown)                                            |
|                                                             |   └ HashJoin[(x = y)] (rows=unknown)                                 |
|                                                             |     ├ Collect[doc.t1 | [x] | true] (rows=unknown)                    |
|                                                             |     └ Filter[(b = 'abc')] (rows=0)                                   |
|                                                             |       └ Collect[doc.t2 | [b, y] | true] (rows=unknown)               |
| optimizer_merge_filter_and_collect                          | Eval[x, b] (rows=unknown)                                            |
|                                                             |   └ HashJoin[(x = y)] (rows=unknown)                                 |
|                                                             |     ├ Collect[doc.t1 | [x] | true] (rows=unknown)                    |
|                                                             |     └ Collect[doc.t2 | [b, y] | (b = 'abc')] (rows=unknown)          |
| Final logical plan                                          | Eval[x, b] (rows=unknown)                                            |
|                                                             |   └ HashJoin[(x = y)] (rows=unknown)                                 |
|                                                             |     ├ Collect[doc.t1 | [x] | true] (rows=unknown)                    |
|                                                             |     └ Collect[doc.t2 | [b, y] | (b = 'abc')] (rows=unknown)          |
+-------------------------------------------------------------+----------------------------------------------------------------------+
EXPLAIN 5 rows in set (3.273 sec)


-- after:
cr> explain verbose select doc.t1.*, doc.t2.b from doc.t1 join doc.t2 on doc.t1.x = doc.t2.y and doc.t2.b = 'abc';                                                                            
+-----------------------------+------------------------------------------------------------+
| STEP                        | QUERY PLAN                                                 |
+-----------------------------+------------------------------------------------------------+
| Initial logical plan        | Eval[x, b] (rows=unknown)                                  |
|                             |   └ Join[INNER | ((x = y) AND (b = 'abc'))] (rows=unknown) |
|                             |     ├ Collect[doc.t1 | [x] | true] (rows=unknown)          |
|                             |     └ Collect[doc.t2 | [b, y] | true] (rows=unknown)       |
| optimizer_rewrite_join_plan | Eval[x, b] (rows=unknown)                                  |
|                             |   └ HashJoin[((x = y) AND (b = 'abc'))] (rows=unknown)     |
|                             |     ├ Collect[doc.t1 | [x] | true] (rows=unknown)          |
|                             |     └ Collect[doc.t2 | [b, y] | true] (rows=unknown)       |
| Final logical plan          | Eval[x, b] (rows=unknown)                                  |
|                             |   └ HashJoin[((x = y) AND (b = 'abc'))] (rows=unknown)     |
|                             |     ├ Collect[doc.t1 | [x] | true] (rows=unknown)          |
|                             |     └ Collect[doc.t2 | [b, y] | true] (rows=unknown)       |
+-----------------------------+------------------------------------------------------------+
```

=> Maybe `optimizer_move_constant_join_conditions_beneath_nested_loop` should be allowed to constant-push-down on hash joins.

## Checklist

 - [ ] Added an entry in the latest `docs/appendices/release-notes/<x.y.0>.rst` for user facing changes
 - [ ] Updated documentation & `sql_features` table for user facing changes
 - [ ] Touched code is covered by tests
 - [ ] [CLA](https://crate.io/community/contribute/cla/) is signed
 - [ ] This does not contain breaking changes, or if it does:
    - It is released within a major release
    - It is recorded in the latest `docs/appendices/release-notes/<x.y.0>.rst`
    - It was marked as deprecated in an earlier release if possible
    - You've thought about the consequences and other components are adapted
      (E.g. AdminUI)

### CODE REVIEW

File: server/src/main/java/io/crate/planner/operators/EquiJoinDetector.java [hunk 1]
```diff
@@ -96,15 +100,13 @@ public Void visitFunction(Function function, Context context) {
...
+                    context.insideEqOperator = saveInsideEqOperator;
                     break;
                 default:
                     if (context.insideEqOperator) {
                         for (Symbol arg : function.arguments()) {
                             arg.accept(this, context);
                         }
-                    } else {
-                        context.isHashJoinPossible = false;
-                        return null;
```
Comment (jeeminso) [inline server/src/main/java/io/crate/planner/operators/EquiJoinDetector.java:107]: Setting `context.isHashJoinPossible` to `true` means an equip-join condition has been observed, there could not be a situation that would revert this decision to my understanding.

**Comment by jeeminso** [comment]
`QuerySplitter` moves the `case` expr from `where` to `join condition` but is not the main cause since a user can directly write such queries.

File: server/src/main/java/io/crate/planner/operators/EquiJoinDetector.java (hunk 1 above)
Comment (mkleen) [reply server/src/main/java/io/crate/planner/operators/EquiJoinDetector.java:107]: what about `a = b AND a + 1 > b` ?

Comment (jeeminso) [reply server/src/main/java/io/crate/planner/operators/EquiJoinDetector.java:107]: Thanks for reviewing. It seems like such join condition is a valid hash join. On current master it does become a hash join. (`HashJoinConditionSymbolsExtractor` chooses the first equality as the hash symbol.)

```
cr> explain verbose select * from t3 join t2 on t3.x = t2.y + 4 and t3.x = t2.y;                                                                                                              
+-----------------------------+----------------------------------------------------------+
| STEP                        | QUERY PLAN                                               |
+-----------------------------+----------------------------------------------------------+
| Initial logical plan        | Join[INNER | ((x = (y + 4)) AND (x = y))] (rows=unknown) |
|                             |   ├ Collect[doc.t3 | [x] | true] (rows=unknown)          |
|                             |   └ Collect[doc.t2 | [y, b] | true] (rows=unknown)       |
| optimizer_rewrite_join_plan | HashJoin[((x = (y + 4)) AND (x = y))] (rows=unknown)     |
|                             |   ├ Collect[doc.t3 | [x] | true] (rows=unknown)          |
|                             |   └ Collect[doc.t2 | [y, b] | true] (rows=unknown)       |
| Final logical plan          | HashJoin[((x = (y + 4)) AND (x = y))] (rows=unknown)     |
|                             |   ├ Collect[doc.t3 | [x] | true] (rows=unknown)          |
|                             |   └ Collect[doc.t2 | [y, b] | true] (rows=unknown)       |
+-----------------------------+----------------------------------------------------------+
EXPLAIN 3 rows in set (0.003 sec)
```

Also looking at the class level javadocs for `EquiJoinDetector` and `HashJoinConditionSymbolsExtractor` I get the impression that we are very lenient in allowing hash joins.

Comment (jeeminso) [reply server/src/main/java/io/crate/planner/operators/EquiJoinDetector.java:107]: I will open a new PR for this sorry for confusion.

**Pull request closed without merging**